### ISSUE TITLE
Files exported by `--file-forwarding` not removed from xdg-document-portal after program exit

### ISSUE THREAD

**Issue by teohhanhui** [flatpak/flatpak#4303]
## Linux distribution and version

openSUSE MicroOS (20210524)

## Flatpak version

Flatpak 1.11.1

## Description of the problem

The documentation for `--file-forwarding` notes that:

> The exports are non-persistent

https://github.com/flatpak/flatpak/blob/686af7d3b81b1315256e79eb5991cede458f3498/doc/flatpak-run.xml#L627-L638

However, often times files exported this way are not removed even after the program started by `flatpak run` has exited. This is causing `target is busy` when trying to unmount a partition on a USB external hard drive.

## Steps to reproduce

1. `flatpak install io.github.celluloid_player.Celluloid`
2. `gio open ???.mkv`
3. Exit the Celluloid player started above.
4. Note the file descriptor for the `???.mkv` video file is still open by `xdg-document-portal`, e.g. by running `lsof /dev/sda1`

**Comment by smcv** [comment]
This might be a file descriptor leak in the `xdg-document-portal`, which is not actually part of Flatpak. Which version of [xdg-desktop-portal](https://github.com/flatpak/xdg-desktop-portal) do you have?

xdg-desktop-portal version 1.8.1 definitely fixed a fd leak, for example.

**Comment by teohhanhui** [comment]
`xdg-desktop-portal` is at version 1.8.1, but I'm definitely still seeing these fd's being left open.

```
teohhanhui@envy-x360:~> sudo pkcon get-packages --filter=installed | grep xdg-desktop-portal
Installed    xdg-desktop-portal-1.8.1-1.2.x86_64 (installed)
Installed    xdg-desktop-portal-gtk-1.8.0-1.6.x86_64 (installed)
```

Do you mean flatpak is correctly requesting the removal of such forwarded files on app exit?

**Comment by teohhanhui** [comment]
I can only find the `xdp_dbus_documents_call_add_sync` call here:

https://github.com/flatpak/flatpak/blob/686af7d3b81b1315256e79eb5991cede458f3498/common/flatpak-run.c#L3335

But there's no `xdp_dbus_documents_call_delete_sync` call on app exit. It seems to only be done in `document-unexport` command:

https://github.com/flatpak/flatpak/blob/686af7d3b81b1315256e79eb5991cede458f3498/app/flatpak-builtins-document-unexport.c#L91

**Comment by rowbawts** [comment]
I'm having the same issue with Clapper when I plug in an SD card and play some videos, close down Clapper and try and eject the SD card gnome tells me volume is busy and running `lsof` on the mounted device it shows xdg-document-portal keeps them open.

**Comment by DaAwesomeP** [comment]
I'm seeing this on openSUSE Leap 15.3 with Flatpak 1.12.5. In my case it is when downloading files in Firefox running in Flatpak.

Are there any known workarounds? Possibly:

- Will temporarily (and manually) adding the location to the Flatpak permissions for the application cause the application to access the file directly instead of using xdg-document-portal?
- Is there a way to force-close or kill xdg-document-portal to cause it to close all of the file descriptors? Or is there a way to simply force those descriptors to close? Does this create a large risk of data loss?

**Comment by hfiguiere** [comment]
Likely https://github.com/flatpak/xdg-desktop-portal/issues/689

**Comment by hfiguiere** [comment]
This should be fixed when the portal with the fix for #689 get released.

**Comment by hfiguiere** [comment]
This should be fixed if you use xdg-desktop-portal 1.18.2.

**Comment by GeorgesStavracas** [comment]
Correct, xdg-desktop-portal 1.18.2 should fix this.
